### PR TITLE
🏥 ♾️ just 4️⃣ 🫵

### PR DESCRIPTION
**Pull request by ishan-renjen**
Added support for STMF446xx GPIO and UART and added functionality so that you can compile for F302x8, F334x8, F446xx depending on compile tag.

### CODE REVIEW

File: include/EVT/io/pin.hpp [hunk 1]
```diff
@@ -101,33 +100,33 @@ enum class Pin {
...
+#endif
+#ifdef STM32F334x8
     SPI_SCK = PA_5, //D13
     SPI_MISO = PA_6,//D12
     SPI_MOSI = PA_7,//D11
     SPI_CS = PB_6,  //D10
-    #endif
+#endif
+
+#ifdef STM32f4xx
```
Comment (mjmagee991) [inline include/EVT/io/pin.hpp:124]: Put this up with the other LED defines

File: include/EVT/io/pin.hpp [hunk 2]
```diff
@@ -8,7 +8,6 @@
 
 namespace EVT::core::IO {
 // STM32F302R8 pin names
```
Comment (mjmagee991) [inline include/EVT/io/pin.hpp:10]:
```suggestion
```

File: include/EVT/io/pin.hpp
Comment (mjmagee991) [inline include/EVT/io/pin.hpp:1]: This file is going to have to change more in the future, but we can leave it as is for now

File: samples/blink/main.cpp
Comment (mjmagee991) [inline samples/blink/main.cpp:1]: Revert all you changes to this file

File: src/dev/RTCTimer.cpp
Comment (mjmagee991) [inline src/dev/RTCTimer.cpp:1]: Revert all your changes to this file

File: src/dev/platform/f3xx/RTCf3xx.cpp
Comment (mjmagee991) [inline src/dev/platform/f3xx/RTCf3xx.cpp:1]: Revert all your changes to this file

File: src/io/CANopen.cpp
Comment (mjmagee991) [inline src/io/CANopen.cpp:1]: Revert all your changes to this file

File: include/EVT/utils/time.hpp [hunk 3]
```diff
@@ -3,13 +3,17 @@
...
  *
  */
-
 #ifndef _EVT_TIME_
 #define _EVT_TIME_
 
 #include <cstdint>
 
-#include <HALf3/stm32f3xx_hal.h>
+#ifdef STM32f4xx
```
Comment (mjmagee991) [inline include/EVT/utils/time.hpp:11]: Instead of doing this, you should break time apart into timef3xx and tim4xx like we have for all the IO stuff

File: include/EVT/utils/types/FixedQueue.hpp [hunk 4]
```diff
@@ -16,7 +16,7 @@
 #ifndef _EVT_FIXED_QUEUE_
 #define _EVT_FIXED_QUEUE_
 
-#include <stdint.h>
+#include <cstddef>
```
Comment (mjmagee991) [inline include/EVT/utils/types/FixedQueue.hpp:19]: Don't bother modifying this in this branch

File: include/EVT/io/platform/f4xx/GPIOf4xx.hpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+    /**
+     * Condenses gpio settings initialization into a single function.
+     * @param targetGpio gpio instance to initialize
+     * @param pins array of pins used by gpio instance
+     * @param numOfPins size of the pin array (either 1 or 2)
+     * @param mode gpio configuration mode
+     * @param pull pull-up or pull-down activation
+     * @param speed maximum gpio output frequency
+     * Possible values for Mode, Pull, and Speed can be found in "stm32f3xx_hal_gpio.h"
```
Comment (mjmagee991) [inline include/EVT/io/platform/f4xx/GPIOf4xx.hpp:46]:
```suggestion
     * Possible values for Mode, Pull, and Speed can be found in "stm32f4xx_hal_gpio.h"
```

File: include/EVT/io/platform/f4xx/GPIOf4xx.hpp [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    GPIOf4xx(Pin pin);
+
+    /**
+     * Create an instance of the STMF4xx GPIO pin using the provided pin
+     * and direction.
+     *
+     * @param[in] pin The pin for the GPIO instance to use.
+     * @param[in] direction The flow of data (either input or output).
+     */
+    GPIOf4xx(Pin pin, Direction direction);
```
Comment (mjmagee991) [inline include/EVT/io/platform/f4xx/GPIOf4xx.hpp:28]: Why did you get rid of pull here?

File: include/EVT/io/platform/f4xx/UARTf4xx.hpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
+#ifndef _EVT_UARTf3xx_
```
Comment (mjmagee991) [inline include/EVT/io/platform/f4xx/UARTf4xx.hpp:1]:
```suggestion
#ifndef _EVT_UARTf4xx_
```

File: include/EVT/io/platform/f4xx/UARTf4xx.hpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+#define _EVT_UARTf3xx_
+
+#include <stdint.h>
```
Comment (mjmagee991) [inline include/EVT/io/platform/f4xx/UARTf4xx.hpp:4]:
```suggestion
#include <cstdint>
```

File: include/EVT/io/platform/f4xx/UARTf4xx.hpp [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+public:
+    /**
+     * Create an instance of the STMF302x8 UART interface using the provided
+     * TX and RX pins.
+     *
+     * @param[in] txPin The UART TX pin.
+     * @param[in] rxPin THe UART RX pin.
+     * @param[in] baudrate The baudrate to operate the UART with/
+     */
+    UARTf4xx(Pin txPin, Pin rxPin, uint32_t baudrate);
```
Comment (mjmagee991) [inline include/EVT/io/platform/f4xx/UARTf4xx.hpp:22]: Why did you get rid of isSwapped?

Comment (mjmagee991) [reply include/EVT/io/platform/f4xx/UARTf4xx.hpp:22]: You probably based this on an old version of UARTf3xx.hpp. Just look at the new version (after the big merge you did) and make it look like that

File: src/io/platform/f4xx/GPIOf4xx.cpp [hunk 10]
```diff
@@ -0,0 +1,213 @@
+#include <EVT/io/platform/f4xx/GPIOf4xx.hpp>
+
+#include <stdint.h>
+
+#include <EVT/io/GPIO.hpp>
+#include <EVT/io/pin.hpp>
+
+#include "EVT/io/platform/f4xx/GPIOf4xx.hpp"
```
Comment (mjmagee991) [inline src/io/platform/f4xx/GPIOf4xx.cpp:8]:
```suggestion
#include <EVT/io/platform/f4xx/GPIOf4xx.hpp>
```

File: src/io/platform/f4xx/GPIOf4xx.cpp [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+
+    GPIO_InitTypeDef gpioInit;
+    Pin myPins[] = {pin};
+    uint8_t numOfPins = 1;
+
+    gpioStateInit(&gpioInit, myPins, numOfPins,
+                  static_cast<uint32_t>(direction), GPIO_PULLDOWN, GPIO_SPEED_FREQ_HIGH);
+
+    switch ((static_cast<uint8_t>(pin) & 0xF0) >> 4) {
+    case 0x0:
```
Comment (mjmagee991) [inline src/io/platform/f4xx/GPIOf4xx.cpp:74]: Add a TODO to double check that this actually applies to F4xx the same way it does to F3xx

File: src/io/platform/f4xx/GPIOf4xx.cpp [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+                  GPIOf4xx::GPIO_TRIGGER_INTERRUPT_BASE | (static_cast<uint32_t>(edge) << GPIO_MODE_IT_SHIFT),
+                  GPIO_PULLDOWN, GPIO_SPEED_FREQ_HIGH);
+
+    auto pin_index = static_cast<uint8_t>(this->pin) & 0x0F;
+    INTERRUPT_HANDLERS[pin_index] = irqHandler;
+    INTERRUPT_GPIOS[pin_index] = this;
+    IRQn_Type irqNum;
+
+    switch (this->halPin) {
+    case GPIO_PIN_0:
```
Comment (mjmagee991) [inline src/io/platform/f4xx/GPIOf4xx.cpp:125]: Add a TODO to check that this applies to F4 the same way it does to F3

File: src/platform/f4xx/stm32f446xx.cpp [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+
+void stm32f4xx_init() {
+    HAL_Init();
+    RCC_OscInitTypeDef RCC_OscInitStruct = {0};
+    RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};
+    RCC_PeriphCLKInitTypeDef PeriphClkInit = {0};
+
+    /** Configure the main internal regulator output voltage
+    */
+    __HAL_RCC_PWR_CLK_ENABLE();
```
Comment (mjmagee991) [inline src/platform/f4xx/stm32f446xx.cpp:19]: I'm pretty sure you're not supposed to call macros that start with `__`. Can you try to find a wrapper function or something to do this instead? Where did you find this code?

File: src/platform/f4xx/stm32f446xx.cpp [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    RCC_OscInitStruct.HSIState = RCC_HSI_ON;
+    RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
+    RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
+    RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSI;
+    RCC_OscInitStruct.PLL.PLLM = 8;
+    RCC_OscInitStruct.PLL.PLLN = 50;
+    RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
+    RCC_OscInitStruct.PLL.PLLQ = 2;
+    RCC_OscInitStruct.PLL.PLLR = 2;
+    HAL_RCC_OscConfig(&RCC_OscInitStruct);
```
Comment (mjmagee991) [inline src/platform/f4xx/stm32f446xx.cpp:35]: Why is this different than the 3xx stuff?

File: src/platform/f4xx/stm32f446xx.cpp [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+
+    /** Initializes the CPU, AHB and APB buses clocks
+    */
+    RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK | RCC_CLOCKTYPE_SYSCLK
+        | RCC_CLOCKTYPE_PCLK1 | RCC_CLOCKTYPE_PCLK2;
+    RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;// SYSCLK at 8 MHz (HSI CLK)
+    RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
+    RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
+    RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;
+    if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_1) != HAL_OK) {
```
Comment (mjmagee991) [inline src/platform/f4xx/stm32f446xx.cpp:49]: Why did you add error handling here?

File: src/platform/f4xx/stm32f446xx.cpp [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+    SysTick_Handler();
+}
+
+void Error_Handler(void) {
+    /* USER CODE BEGIN Error_Handler_Debug */
+    /* User can add his own implementation to report the HAL error return state */
+    __disable_irq();
+    while (1) {
+        std::cout << "the code died" << endl;
```
Comment (mjmagee991) [inline src/platform/f4xx/stm32f446xx.cpp:70]: This won't do anything. We don't have `cout` on a microcontroller

File: src/utils/platform/f4xx/time.cpp
Comment (mjmagee991) [inline src/utils/platform/f4xx/time.cpp:1]: As I said somewhere else, this needs to be renamed and set up as timef4xx.cpp

File: libs/HALf4/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+install(
+    EXPORT ${PROJECT_NAME}-config
+    NAMESPACE ${PROJECT_NAME}::
+    DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
+)
+
+install(
+    DIRECTORY ${CMAKE_CURRENT_LIST_DIR}/include/
+    DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME}
+)
```
Comment (mjmagee991) [inline libs/HALf4/CMakeLists.txt:81]:
```suggestion
)

```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
You've done an awesome job, but there's more to be done before we can call this one finished

File: cmake/evt-core_build.cmake [hunk 18]
```diff
@@ -5,7 +5,6 @@ files for the given name. This will make a dedicated project.
 macro(make_exe proj_name sources)
     project(${proj_name} C CXX ASM)
     add_definitions(-DUSE_HAL_LIBRARY)
-
```
[the remark this answers is not in the document]
Comment (ishan-renjen) [reply cmake/evt-core_build.cmake:8]: i didnt change any code here

File: include/EVT/platform/f3xx/stm32f3xx.hpp [hunk 19]
```diff
@@ -1,6 +1,8 @@
 #ifndef _EVT_STM32F3xx_
 #define _EVT_STM32F3xx_
 
+#include <cstdint>
```
[the remark this answers is not in the document]
Comment (ishan-renjen) [reply include/EVT/platform/f3xx/stm32f3xx.hpp:4]: if thats not there the uint32 constants arent happy

File: include/EVT/platform/f4xx/stm32f4xx.hpp [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+constexpr uint32_t TIMER_INTERRUPT_PRIORITY = 9;
+constexpr uint32_t GPIO_INTERRUPT_PRIORITY = 10;
+
+/**
+ * Handles system level initialization of the STM32F4xx. This makes a
+ * series of calls into the HAL to enable system peripherals and enable
+ * all required clock.
+ */
+void stm32f4xx_init();
+void Error_Handler(void);
```
[the remark this answers is not in the document]
Comment (ishan-renjen) [reply include/EVT/platform/f4xx/stm32f4xx.hpp:21]: because i did not modify the f3 code

File: include/EVT/io/platform/f4xx/ADCf4xx.hpp
[the remark this answers is not in the document]
Comment (ishan-renjen) [reply include/EVT/io/platform/f4xx/ADCf4xx.hpp:1]: I didnt do that i think whoever started the f4 stuff did

File: src/platform/f4xx/stm32f446xx.cpp (hunk 14 above)
Comment (ishan-renjen) [reply src/platform/f4xx/stm32f446xx.cpp:35]: bc there are different scalars for the pll clock on the f4's than the f3's

File: src/platform/f4xx/stm32f446xx.cpp (hunk 15 above)
Comment (ishan-renjen) [reply src/platform/f4xx/stm32f446xx.cpp:49]: because the autogenerated code did

File: include/EVT/manager.hpp [hunk 21]
```diff
@@ -7,195 +7,260 @@
...
- */
+* Get an instance of an IWDG
+*
+* @param ms Time in milliseconds before the IWDG triggers a reset
+* must be a value between 8 and 32768 ms.
+*/
+#ifdef IWDG_SUPPORTED
 IWDG& getIWDG(uint32_t ms) {
-#ifdef STM32F3xx
+    #ifdef STM32F3xx
```
Comment (mjmagee991) [inline include/EVT/manager.hpp:82]: No macros should be indented, even though they're technically nested

File: include/EVT/manager.hpp [hunk 22]
```diff
@@ -7,195 +7,260 @@
...
+    #endif
 }
+#endif
 
 /**
- * Get an instance of a PWM pin.
- *
- * @param[in] pin The pin to attach to the PWM.
- */
+* Get an instance of a PWM pin.
```
Comment (mjmagee991) [inline include/EVT/manager.hpp:194]: Why did you de-dent all these comments? The `*` should be in line with the first `*` on the first line

File: samples/blink/main.cpp
Comment (mjmagee991) [inline samples/blink/main.cpp:1]: Why did you modify this file?

File: src/platform/f4xx/stm32f446xx.cpp (hunk 13 above)
Comment (mjmagee991) [reply src/platform/f4xx/stm32f446xx.cpp:19]: Did you find a resolution for this?

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
A couple more questions. Let's talk today

File: include/EVT/dev/LCD.hpp [hunk 23]
```diff
@@ -5,7 +5,7 @@
 #include <EVT/io/SPI.hpp>
 #include <EVT/io/UART.hpp>
 #include <EVT/io/pin.hpp>
-#include <EVT/utils/time.hpp>
+//#include <EVT/utils/time.hpp>
```
[the remark this answers is not in the document]
Comment (ishan-renjen) [reply include/EVT/dev/LCD.hpp:8]: yea it builds

File: samples/blink/main.cpp
Comment (ishan-renjen) [reply samples/blink/main.cpp:1]: you told me to put it back to normal so i copied it from evt-core on github

File: src/platform/f4xx/stm32f446xx.cpp (hunk 13 above)
Comment (ishan-renjen) [reply src/platform/f4xx/stm32f446xx.cpp:19]: no, i dont want to touch code that was already there and works

Comment (ishan-renjen) [reply src/platform/f4xx/stm32f446xx.cpp:19]: although it is interesting that the comment above is wrong

**Review by mjmagee991 — APPROVED** [review]
Pog
